javascript/load-gen.ts

```diff
@@ -0,0 +1,406 @@
+    this.totalNumberOfOperationsToExecute =
+      options.totalNumberOfOperationsToExecute;
+
+    this.cacheValue = 'x'.repeat(options.cacheItemPayloadBytes);
+  }
+
+  async run(): Promise<void> {
+    const momento = new SimpleCacheClient(
+      this.authToken,
+      this.cacheItemTtlSeconds,
+      {
+        requestTimeoutMs: this.requestTimeoutMs,
+        loggerOptions: this.loggerOptions,
+      }
+    );
+
+    try {
+      await momento.createCache(this.cacheName);
+    } catch (e) {
+      if (e instanceof AlreadyExistsError) {
+        this.logger.info('cache already exists');
+      } else {
+        throw e;
+      }
+    }
+
+    const numRequestsPerWorker =
+      this.totalNumberOfOperationsToExecute / this.numberOfConcurrentRequests;
+
+    const loadGenContext: BasicJavasScriptLoadGenContext = {
+      startTime: process.hrtime(),
+      getLatencies: hdr.build(),
+      setLatencies: hdr.build(),
+      globalRequestCount: 0,
+      globalSuccessCount: 0,
+      globalUnavailableCount: 0,
+      globalDeadlineExceededCount: 0,
+      globalResourceExhaustedCount: 0,
+      globalRstStreamCount: 0,
+    };
+
+    const asyncGetSetResults = range(this.numberOfConcurrentRequests).map(
+      workerId =>
+        this.launchAndRunWorkers(
+          momento,
+          loadGenContext,
+          workerId + 1,
+          numRequestsPerWorker
+        )
+    );
+    const allResultPromises = Promise.all(asyncGetSetResults);
+    await allResultPromises;
+    this.logger.info('DONE!');
+    // wait a few millis to allow the logger to finish flushing
+    await delay(500);
+  }
+
+  private async launchAndRunWorkers(
+    client: SimpleCacheClient,
+    loadGenContext: BasicJavasScriptLoadGenContext,
+    workerId: number,
+    numRequests: number
+  ): Promise<number> {
+    let successCount = 0;
+    for (let i = 1; i <= numRequests; i++) {
+      const result = await this.issueAsyncSetGet(
+        client,
+        loadGenContext,
+        workerId,
+        i
+      );
+      loadGenContext.globalRequestCount++;
+      // TODO: this could be simplified and made more generic, worth doing if we ever want to
+      //  expand this to additional types of behavior
+      switch (result) {
+        case AsyncSetGetResult.SUCCESS:
+          successCount++;
+          loadGenContext.globalSuccessCount++;
+          break;
+        case AsyncSetGetResult.UNAVAILABLE:
+          loadGenContext.globalUnavailableCount++;
+          break;
+        case AsyncSetGetResult.DEADLINE_EXCEEDED:
+          loadGenContext.globalDeadlineExceededCount++;
+          break;
+        case AsyncSetGetResult.RESOURCE_EXHAUSTED:
+          loadGenContext.globalResourceExhaustedCount++;
+          break;
+        case AsyncSetGetResult.RST_STREAM:
+          loadGenContext.globalRstStreamCount++;
+          break;
+      }
+      if (loadGenContext.globalRequestCount % 1000 === 0) {
```

Comment:
I think you want to reset (several of) the statistics after printing right? Otherwise, for example, the p999 will track more and more "optimistic" requests as the test runs on

Reply:
IMO we can get more fancy with this in the future, but right now this test only runs for < 5 mins and I'm really just interested in the data over the course of the whole run.  we can discuss further if you disagree.

Reply:
i mean okay but

```typescript
loadGenContext.setLatencies.reset()
loadGenContext.getLatencies.reset()
```

at the end of this output block neither seems terribly fancy nor too difficult to consider for the benefit of more faithful reporting per interval imho

here, see it works: https://runkit.com/embed/2639xw5hwkak

as-is, the very meaning of this data changes (mathematical value decreases) with every subsequent interval. Would be nice to consider doing this.

Reply:
i'm not suggesting that it's difficult, i'm suggesting that for a run that is only 5 minutes long, my instinct is that i want stats for the whole run rather than 50 sets of stats for different small intervals during the run.  it's not the same thing that I would want for an operational dashboard obviously, but for this tool I think users are just going to look at the last couple of printouts on the console and I don't know if it's good/faithful for those to not include any of the data from earlier in the run?

Reply:
50 printouts of intermediate data seems misleading then, and seems like so do you:
> I think users are just going to look at the last couple of printouts on the console

but only the very last one is showing the data you're interested in. Every prior printout shows a smaller window toward the very first window, which has maximum detail / maximum pessimism.

> my instinct is that i want stats for the whole run

when you put it this way it makes a lot of sense, I don't know why I didn't understand that's what you were after before. So maybe the %1000 printout should only contain some kind of `progress: ${(i / totalIterations * 100).toFixed(1)}%` and the end printout should be showing the stats you want?

Reply:
Added the word 'cumulative' to the stats output in d423998132f04e68fb33060da3821826b95fc37c